bounded_async_foreach_test.cpp

```diff
@@ -205,3 +204,40 @@ TEST(bounded_async_foreach, synchronous_calls_test) {
   EXPECT_TRUE(finish_called);
   EXPECT_EQ(done_invocations, input);
 }
+
+TEST(bounded_async_foreach, operation_canceled) {
+
+    asio::io_context ctx;
+
+    bool finish_called = false;
+    std::vector<int> input = { 1, 2, 3, 4, 5 };
+    std::vector<int> done_invocations;
+    bounded_async_foreach(
+        2, input,
+        [&](auto element, auto done) {
+            EXPECT_FALSE(finish_called) << element;
+            done_invocations.emplace_back(element);
+            auto result = std::error_code();
+            auto timeout = std::chrono::milliseconds(10);
+            if (element == 1) {
+                result = std::make_error_code(std::errc::operation_canceled);
+                timeout = std::chrono::milliseconds(5);
+            }
+            async_sleep(ctx, timeout, [=] {
+                done(result);
+            });
+        },
+        [&](std::error_code ec) {
+            EXPECT_TRUE(ec);
+            finish_called = true;
+        });
+
+    EXPECT_FALSE(finish_called);
+    EXPECT_EQ(done_invocations.size(), 0);
```

Comment:
this expectation seems to be wrong. it needs to call callbacks while it has free "workers" so it will 2.